metadata/src/main/java/org/apache/kafka/controller/AclControlManager.java

```diff
@@ -0,0 +1,232 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.kafka.controller;
+
+import org.apache.kafka.common.Uuid;
+import org.apache.kafka.common.acl.AclBinding;
+import org.apache.kafka.common.acl.AclBindingFilter;
+import org.apache.kafka.common.errors.ApiException;
+import org.apache.kafka.common.errors.InvalidRequestException;
+import org.apache.kafka.common.errors.UnknownServerException;
+import org.apache.kafka.common.metadata.AccessControlEntryRecord;
+import org.apache.kafka.common.metadata.RemoveAccessControlEntryRecord;
+import org.apache.kafka.common.requests.ApiError;
+import org.apache.kafka.metadata.authorizer.ClusterMetadataAuthorizer;
+import org.apache.kafka.metadata.authorizer.StandardAcl;
+import org.apache.kafka.metadata.authorizer.StandardAclRecordIterator;
+import org.apache.kafka.metadata.authorizer.StandardAclWithId;
+import org.apache.kafka.raft.OffsetAndEpoch;
+import org.apache.kafka.server.authorizer.AclCreateResult;
+import org.apache.kafka.server.authorizer.AclDeleteResult;
+import org.apache.kafka.server.authorizer.AclDeleteResult.AclBindingDeleteResult;
+import org.apache.kafka.server.common.ApiMessageAndVersion;
+import org.apache.kafka.timeline.SnapshotRegistry;
+import org.apache.kafka.timeline.TimelineHashMap;
+import org.apache.kafka.timeline.TimelineHashSet;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import java.util.Map.Entry;
+import java.util.Optional;
+
+
+/**
+ * The AclControlManager manages any ACLs that are stored in the __cluster_metadata topic.
+ * If the ACLs are stored externally (such as in ZooKeeper) then there will be nothing for
+ * this manager to do, and the authorizer field will always be Optional.empty.
+ *
+ * Because the Authorizer is being concurrently used by other threads, we need to be
+ * careful about snapshots. We don't want the Authorizer to act based on partial state
+ * during the loading process. Therefore, unlike most of the other managers,
+ * AclControlManager needs to receive callbacks when we start loading a snapshot and when
+ * we finish. The prepareForSnapshotLoad callback clears the authorizer field, preventing
+ * any changes from affecting the authorizer until completeSnapshotLoad is called.
+ * Note that the Authorizer's start() method will block until the first snapshot load has
+ * completed, which is another reason the prepare / complete callbacks are needed.
+ */
+public class AclControlManager {
+    private final TimelineHashMap<Uuid, StandardAcl> idToAcl;
+    private final TimelineHashSet<StandardAcl> existingAcls;
+    private final Optional<ClusterMetadataAuthorizer> authorizer;
```

Comment:
If there is no authorizer configured, or a non-cluster-metadata one.

Reply:
I guess my expectation was that we would only construct `AclControlManager` if there was a `ClusterMetadataAuthorizer` implementation. Why would we need it otherwise?

Reply:
We don't need it, but the extra memory it takes up is insignificant, and we don't want to have to test multiple code paths.